copycat/api/src/main/java/org/apache/kafka/copycat/data/SchemaProjector.java

```diff
@@ -0,0 +1,209 @@
+            throw new SchemaProjectorException("Key schema type mismatch for array. Writer key schema type: " + writer.type() + " and Reader key schema type: " + reader.type());
+        }
+        Type writerValueType = writer.valueSchema().type();
+        Type readerValueType = reader.valueSchema().type();
+        if (writerValueType != readerValueType && !promote(writerValueType, readerValueType)) {
+            throw new SchemaProjectorException("Value schema type mismatch for array. Writer value schema type: " + writer.type() + " and Reader value schema type: " + reader.type());
+        }
+        Map<?, ?> map = (Map<?, ?>) record;
+        Map<Object, Object> retMap = new HashMap<>();
+        for (Map.Entry<?, ?> entry : map.entrySet()) {
+            Object key = entry.getKey();
+            Object value = entry.getValue();
+            Object retKey = project(writer.keySchema(), reader.keySchema(), key);
+            Object retValue = project(writer.valueSchema(), writer.valueSchema(), value);
+            retMap.put(retKey, retValue);
+        }
+        return retMap;
+    }
+
+    private static Object projectLogical(Schema writer, Schema reader, Object record) throws SchemaProjectorException {
+        if (writer.name() == null || !writer.name().equals(reader.name())) {
+            throw new SchemaProjectorException("Schema does not match for " + writer.name() + " and " + reader.name());
+        }
+
+        switch (reader.name()) {
+            case Decimal.LOGICAL_NAME:
+            case Date.LOGICAL_NAME:
+            case Time.LOGICAL_NAME:
+            case Timestamp.LOGICAL_NAME:
+                return record;
+            default:
+                throw new SchemaProjectorException("Not a valid logical type: " + reader.name());
+        }
+    }
+
+    private static Object projectPrimitive(Schema writer, Schema reader, Object record) throws SchemaProjectorException {
+        if (writer.type() != reader.type() && !promote(writer.type(), reader.type())) {
+            throw new SchemaProjectorException("Schema type mismatch. Writer type: " + writer.type() + " and Reader type: " + reader.type());
+        }
+        if (writer.isOptional() && !reader.isOptional()) {
+            if (reader.defaultValue() != null) {
+                if (record != null) {
+                    return record;
+                } else {
+                    return reader.defaultValue();
+                }
+            } else {
+                throw new SchemaProjectorException("Writer schema is optional, however, reader schema does not provide a default value.");
+            }
+        } else {
+            return record;
+        }
+    }
+
+    private static boolean promote(Type writerType, Type readerType) {
+        Type[] numericTypes = {Type.INT8, Type.INT16, Type.INT32, Type.INT64, Type.FLOAT32, Type.FLOAT64};
+        int index = -1;
+        for (int i = 0; i < numericTypes.length; ++i) {
```

Comment:
This list is small anyway, so I'm not too worried about the cost, but this seems wasteful. Seems like this promotion check should be a simple lookup in a precomputed table.